### PR TITLE
Add array_overlap function

### LINKED ISSUE #6823: Array Intersect Exists
Build a function:
array_intersect_exists(array<E>, array<E>) -> boolean

That basically checks if the intersect of the two input arrays is NOT NULL. This would be more optimized than doing a cardinality(array_intersect(a1,a2)) > 0, since this would exit out once a common element is found.

### PR DESCRIPTION
**Pull request by idemura**
Checked and unchecked (for sorting) version is like 11 vs 19 seconds (walltime). I think compared to ~1.5-2 minutes it's still a win. They have very poor parallelization of this kind of a query (just 6 nodes for 100M rows). 

Fixes #6823

### CODE REVIEW

File: presto-docs/src/main/sphinx/functions/array.rst [hunk 1]
```diff
@@ -37,6 +37,10 @@ Array Functions
 
     Concatenates the elements of the given array using the delimiter and an optional string to replace nulls.
 
+.. function:: arrays_overlap(x, y) -> boolean
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/array.rst]: Function names should be sorted, so move before `cardinality` .

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        int leftPosition = 0;
+        int rightPosition = 0;
+        while (leftPosition < leftSize && rightPosition < rightSize) {
+            if (arrayIntLeft[leftPosition] < arrayIntRight[rightPosition]) {
+                leftPosition++;
+            }
+            else if (arrayIntLeft[leftPosition] > arrayIntRight[rightPosition]) {
+                rightPosition++;
+            }
+            else {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Else is redundant here

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    private long[] arrayBigIntRight;
+    private int[] arrayIntLeft;
+    private int[] arrayIntRight;
+
+    @TypeParameter("E")
+    public ArraysOverlapFunction(@TypeParameter("E") Type elementType)
+    {
+    }
+
+    public ArraysOverlapFunction()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Why is this needed?

File: presto-docs/src/main/sphinx/functions/array.rst [hunk 4]
```diff
@@ -37,6 +37,10 @@ Array Functions
...
+
+    Returns true if two arrays have elements in common. Arrays must contain no NULLs.
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/array.rst]:
```
Returns true if arrays ``x`` and ``y`` have elements in common.
The arrays must not contain nulls.

Comment (electrum) [inline presto-docs/src/main/sphinx/functions/array.rst]: I don't understand the null comment. The code seems to handle nulls fine.

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 5]
```diff
@@ -573,6 +574,37 @@ public void testSlice()
     }
 
     @Test
+    public void testArrayIntersectCompare()
+            throws Exception
+    {
+        assertFunction("ARRAYS_OVERLAP(ARRAY [1, 2], ARRAY [2, 3])", BooleanType.BOOLEAN, true);
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Add tests for arrays with nulls

**Comment by electrum** [comment]
The commit message should be "Add arrays_overlap function"

**Comment by electrum** [comment]
There is a Travis failure

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+import it.unimi.dsi.fastutil.ints.AbstractIntComparator;
+import it.unimi.dsi.fastutil.ints.IntArrays;
+import it.unimi.dsi.fastutil.ints.IntComparator;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Arrays;
+
+import static com.facebook.presto.spi.function.OperatorType.LESS_THAN;
+
+@ScalarFunction("arrays_overlap")
```
Comment (geraint0923) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: `array_overlap`?

**Comment by electrum** [comment]
We might be able to make this faster for the sorted case by doing a first pass to simultaneously check intersection and sorted. If there's an intersection, return true immediately. If not sorted, break and sort.

For the case of sorted and early matches, this could be better than O(n). However, this makes the loop more complex, so it could be worse for other cases.

It'd be really helpful to have a JMH benchmark.

**Comment by idemura** [comment]
I think simultaneous processing adds lots of complexity. Second pass on same data, which is in cache - very cheap. You save on int compare and increment, this is tiny.

Here is big confusion about big O. Big O means "longest time to run algorithm". In it will be O(N) no matter what (in this strategy). You save on average (theta) time. It's not that big difference compared to avoiding indirected sort and access + devirtualization of comparison. Other factor are tiny according to experiments.

**Comment by idemura** [comment]
Updated. PTAL. Handles nulls.

**Comment by haozhun** [comment]
Minor comment. I notice that you renamed the commit to `Add array_overlap_function`. It looks like @electrum asked for `Add array_overlap function`. I think the latter is better.

**Comment by haozhun** [comment]
In addition, it looks like you didn't assign this pull request to any body. The reviewers likely won't notice this pull request unless you assign it to them.

**Comment by martint** [comment]
arrays_overlap is good for the name

**Comment by idemura** [comment]
Updated. I initially set reviewer as I remember.

**Comment by haozhun** [comment]
Reviewer is not sufficient. You have to set assignee. GitHub does not let you find everything pending your review.

**Comment by idemura** [comment]
@electrum please take a look if your requests are satisfied.

**Comment by electrum** [comment]
Please update the commit message to match the function name

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 3 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: This seems suspicious. @haozhun do you know what this is about?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    private int[] leftPositions = new int[INITIAL_SIZE];
+    private int[] rightPositions = new int[INITIAL_SIZE];
+
+    private long[] arrayBigIntLeft;
+    private long[] arrayBigIntRight;
+    private int[] arrayIntLeft;
+    private int[] arrayIntRight;
+
+    @TypeParameter("E")
+    public ArraysOverlapFunction(@TypeParameter("E") Type elementType)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Use `{}` on same line for intentionally empty constructors

Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Done

**Comment by idemura** [comment]
Comment updated.

**Comment by idemura** [comment]
@haozhun please take final pass

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+            }
+        }
+        boolean leftWithNull = leftArray.isNull(leftPositions[leftPositionCount - 1]);
+        boolean rightWithNull = rightArray.isNull(rightPositions[rightPositionCount - 1]);
+        return leftWithNull || rightWithNull ? null : false;
+    }
+
+    private static void sortIfUnsorted(long[] array, int size)
+    {
+        if (size == 1) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: This `if` is unnecessary. I think it's unlikely to provide visible performance benefits. And it's confusing that it handles 1 but not zero.

File: presto-docs/src/main/sphinx/functions/array.rst [hunk 9]
```diff
@@ -58,6 +58,10 @@ Array Functions
     Sorts and returns the array ``x``. The elements of ``x`` must be orderable.
     Null elements will be placed at the end of the returned array.
 
+.. function:: arrays_overlap(x, y) -> boolean
+
+    Returns true if arrays ``x`` and ``y`` have elements in common.
```
Comment (haozhun) [inline presto-docs/src/main/sphinx/functions/array.rst]: I think it's worthwhile to document the `null` behavior here.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    {
+        int leftPositionCount = leftArray.getPositionCount();
+        int rightPositionCount = rightArray.getPositionCount();
+
+        if (leftPositionCount == 0 || rightPositionCount == 0) {
+            return false;
+        }
+
+        if (leftPositions.length < leftPositionCount) {
+            leftPositions = new int[leftPositionCount];
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: This can result in degenerative behavior. See my comments for allocating arrays in the other functions.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+                rightCurrentPosition++;
+            }
+            else if (compareValue < 0) {
+                leftCurrentPosition++;
+            }
+            else {
+                return true;
+            }
+        }
+        boolean leftWithNull = leftArray.isNull(leftPositions[leftPositionCount - 1]);
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Add comment that your comparator puts nulls last.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+import static com.facebook.presto.spi.function.OperatorType.LESS_THAN;
+
+@ScalarFunction("arrays_overlap")
+@Description("Returns true if arrays have common elements")
+public final class ArraysOverlapFunction
+{
+    private static final int INITIAL_SIZE = 128;
+
+    private int[] leftPositions = new int[INITIAL_SIZE];
+    private int[] rightPositions = new int[INITIAL_SIZE];
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: This is code came from ArraysIntersectFunction. Not initialization was my idea (and size it * 2)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+    private long[] arrayBigIntLeft;
+    private long[] arrayBigIntRight;
+    private int[] arrayIntLeft;
+    private int[] arrayIntRight;
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+                return type.compareTo(block, left, block, right);
+            }
+        };
+    }
+
+    @SqlNullable
+    @SqlType(StandardTypes.BOOLEAN)
+    public Boolean arraysOverlapInt(
+            @OperatorDependency(operator = LESS_THAN, returnType = StandardTypes.BOOLEAN, argumentTypes = {"integer", "integer"}) MethodHandle lessThanFunction,
+            @SqlType("array(integer)") Block leftArray,
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: No, there shouldn't problems like this. I didn't know that AbstractLongType can extraction for me. It reduces code significantly.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+    {
+        int leftSize = leftArray.getPositionCount();
+        int rightSize = rightArray.getPositionCount();
+
+        if (leftSize == 0 || rightSize == 0) {
+            return false;
+        }
+
+        if (arrayIntLeft == null || arrayIntLeft.length < leftSize) {
+            arrayIntLeft = new int[leftSize * 2];
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: No, why? we take max (which is leftSize) times two.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 14 above)
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Oh, actually int and long are different base, I need an accessor.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+        }
+        if (arrayIntRight == null || arrayIntRight.length < rightSize) {
+            arrayIntRight = new int[rightSize * 2];
+        }
+
+        boolean leftWithNull = false;
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+        int leftRealSize = 0;
+        for (int i = 0; i < leftSize; i++) {
+            if (leftArray.isNull(i)) {
+                leftWithNull = true;
+            }
+            else {
+                arrayIntLeft[leftRealSize++] = leftArray.getInt(i, 0);
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Code generalized now

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+        }
+
+        boolean leftWithNull = false;
+        int leftRealSize = 0;
+        for (int i = 0; i < leftSize; i++) {
+            if (leftArray.isNull(i)) {
+                leftWithNull = true;
+            }
+            else {
+                arrayBigIntLeft[leftRealSize++] = leftArray.getLong(i, 0);
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Ditto

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+        sortIfUnsorted(arrayBigIntLeft, leftRealSize);
+
+        boolean rightWithNull = false;
+        int rightRealSize = 0;
+        for (int i = 0; i < rightSize; i++) {
+            if (rightArray.isNull(i)) {
+                rightWithNull = true;
+            }
+            else {
+                arrayBigIntRight[rightRealSize++] = rightArray.getLong(i, 0);
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Ditto

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 8 above)
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+            if (array[i - 1] > array[i]) {
+                Arrays.sort(array, 0, size);
+                return;
+            }
+        }
+    }
+
+    private static void sortIfUnsorted(int[] array, int size)
+    {
+        if (size == 1) {
```
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Ditto

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 10 above)
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: max? it should be fine

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 11 above)
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: This is a comparator we usually have for similar functions, I don't think we need to comment this additionally only here.

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 21]
```diff
@@ -573,6 +574,61 @@ public void testSlice()
...
+        assertFunction("ARRAYS_OVERLAP(ARRAY [CAST(1 AS BIGINT), 2], ARRAY [CAST(3 AS BIGINT), 2])", BooleanType.BOOLEAN, true);
+
+        assertFunction("ARRAYS_OVERLAP(ARRAY [CAST(1 AS BIGINT), 3], ARRAY [CAST(2 AS BIGINT), 4])", BooleanType.BOOLEAN, false);
+        assertFunction("ARRAYS_OVERLAP(ARRAY [CAST(3 AS BIGINT), 1], ARRAY [CAST(2 AS BIGINT), 4])", BooleanType.BOOLEAN, false);
+        assertFunction("ARRAYS_OVERLAP(ARRAY [CAST(3 AS BIGINT), 1], ARRAY [CAST(4 AS BIGINT), 2])", BooleanType.BOOLEAN, false);
+        assertFunction("ARRAYS_OVERLAP(ARRAY [CAST(1 AS BIGINT), 3], ARRAY [CAST(4 AS BIGINT), 2])", BooleanType.BOOLEAN, false);
+
+        assertFunction("ARRAYS_OVERLAP(ARRAY ['dog', 'cat'], ARRAY ['monkey', 'dog'])", BooleanType.BOOLEAN, true);
+        assertFunction("ARRAYS_OVERLAP(ARRAY ['dog', 'cat'], ARRAY ['monkey', 'fox'])", BooleanType.BOOLEAN, false);
+
```
Comment (idemura) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Done

**Comment by idemura** [comment]
PTAL

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+    private static int sortAbstractLongArray(Block array, long[] buffer, AbstractType type)
+    {
+        int arraySize = array.getPositionCount();
+        int nonNullSize = 0;
+        for (int i = 0; i < arraySize; i++) {
+            if (!array.isNull(i)) {
+                buffer[nonNullSize++] = type.getLong(array, i);
+            }
+        }
+        for (int i = 1; i < nonNullSize; i++) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java]: Add comment "// sort if not already sorted".

File: presto-docs/src/main/sphinx/functions/array.rst [hunk 23]
```diff
@@ -58,6 +58,12 @@ Array Functions
     Sorts and returns the array ``x``. The elements of ``x`` must be orderable.
     Null elements will be placed at the end of the returned array.
 
+.. function:: arrays_overlap(x, y) -> boolean
+
+    Returns true if arrays ``x`` and ``y`` have non-null elements in common.
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/array.rst]:
```
    Tests if arrays ``x`` and ``y`` have any any non-null elements in common.
    Returns null if there are no non-null elements in common but either array contains null.

Comment (idemura) [inline presto-docs/src/main/sphinx/functions/array.rst]: I like this wording. Thanks.
